File: 2-DataStructures/2-Linked-Lists/2-DoublyLinkedLists/doubly-linked-lists.js
// Doubly-linked lists (definition): a linked list that has a pointer to the previous node
// Doubly-linked lists (pros): can be iterated in both directions
// Doubly-linked lists (cons): more memory, more complex code

class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
    this.prev = null; // new
  }
}

class DoublyLinkedList {
  constructor(value) {
    this.head = new Node(value); // the head and tail are the same node
    this.tail = this.head;
    this.length = 1;
  }

  append(value) { // O(1)
    const newNode = new Node(value);
    newNode.prev = this.tail; // new
    this.tail.next = newNode;
    this.tail = newNode;
    this.length++;
    return this;
  }

  prepend(value) { // O(1)
    const newNode = new Node(value);
    newNode.next = this.head;
    this.head = newNode;
    this.head.prev = newNode; // new
    this.length++;
    return this;
  }

  insert(index, value) { // O(n)
    if (index >= this.length) {
      return this.append(value);
    }
    if (index === 0) {
      return this.prepend(value);
    }
    const newNode = new Node(value);
    const leader = this.traverseToIndex(index - 1);
    const follower = leader.next;
    leader.next = newNode;
    newNode.prev = leader;
    newNode.next = follower;
    follower.prev = newNode;
    this.length++;
    return this;
  }

  traverseToIndex(index) { // O(n)
    let counter = 0;
    let currentNode = this.head;
    while (counter !== index) {
      currentNode = currentNode.next;
      counter++;
    }
    return currentNode;
  }

  print() {
    const array = [];
    let currentNode = this.head;
    while (currentNode !== null) {
      array.push(currentNode.value)
      currentNode = currentNode.next;
    }
    console.log(array);
  }

  remove(index) { // O(n)
    if (index === 0) {
      this.head = this.head.next;
      this.length--;
      return this;
    }
    if (index >= this.length) {
      index = this.length - 1;
    }
    const leader = this.traverseToIndex(index - 1);
    const unwantedNode = leader.next;
    leader.next = unwantedNode.next;
    leader.prev = unwantedNode.prev; // new
    this.length--;
    return this;
  }
}

const myDoublyLinkedList = new DoublyLinkedList(10);
myDoublyLinkedList.print();
myDoublyLinkedList.append(5);
myDoublyLinkedList.append(16);
myDoublyLinkedList.print();
myDoublyLinkedList.prepend(1);
myDoublyLinkedList.print();
myDoublyLinkedList.insert(2, 99);
myDoublyLinkedList.print();
myDoublyLinkedList.remove(2);
myDoublyLinkedList.print();

File: 2-DataStructures/2-Linked-Lists/2-DoublyLinkedLists/doubly-linked-lists.py
# doubly-linked lists definition: it is a list that has a head and a tail
# and each node has a pointer to the next node and the previous node
# the tail's next node is None and the head's previous node is None
# the head's previous node is None and the tail's next node is None

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    # add a node to the end of the list, so we have to manipulate the tail node in O(1) runtime
    def insert(self, value):
        newNode = Node(value)
        if self.head is None:
            self.head = newNode
            self.tail = newNode
        else:
            # the tail's next node is the new node
            self.tail.next = newNode
            # the new node's previous node is the tail
            newNode.prev = self.tail
            # the new node is the new tail
            self.tail = newNode
        self.length += 1
        return self

    # traversing the double linked lists in both directions
    def traverse_forward(self):
        current = self.head
        while current is not None:
            print(current.value)
            current = current.next

    def traverse_backward(self):
        current = self.tail
        while current is not None:
            print(current.value)
            current = current.prev


if __name__ == "__main__":
    myDoublyLinkedList = DoublyLinkedList()
    myDoublyLinkedList.insert(1)
    myDoublyLinkedList.insert(2)
    myDoublyLinkedList.insert(3)
    myDoublyLinkedList.insert(4)
    myDoublyLinkedList.insert(5)
    myDoublyLinkedList.insert(6)
    myDoublyLinkedList.insert(7)
    myDoublyLinkedList.insert(8)
    myDoublyLinkedList.insert(9)
    myDoublyLinkedList.insert(10)
    myDoublyLinkedList.traverse_forward()
    print("**********")
    myDoublyLinkedList.traverse_backward()
